src/swap/components/SwapProvider.tsx

```diff
@@ -184,21 +184,74 @@ export function SwapProvider({
           return handleError({ swapError: response });
         }
 
-        onSubmit?.(response);
+        const { transaction, approveTransaction } = response;
+
+        if (approveTransaction?.data) {
+          setPendingTransaction(true);
+          const approveTxHash = await sendTransactionAsync({
+            to: approveTransaction.to,
+            value: approveTransaction.value,
+            data: approveTransaction.data,
+          });
+          await waitForTransactionReceipt(config, {
+            hash: approveTxHash,
+            confirmations: 1,
+          });
+          setPendingTransaction(false);
+        }
+
+        setPendingTransaction(true);
+        const txHash = await sendTransactionAsync({
+          to: transaction.to,
+          value: transaction.value,
+          data: transaction.data,
+        });
+        setPendingTransaction(false);
+
+        setLoading(true);
+        await waitForTransactionReceipt(config, {
+          hash: txHash,
+          confirmations: 1,
+        });
+
+        // TODO: refresh balances
       } catch (e) {
-        handleError({ swapError: e as SwapError });
+        const userRejected = (e as BaseError).message.includes(
+          'User rejected the request.',
+        );
+        if (userRejected) {
+          setLoading(false);
+          setPendingTransaction(false);
+          handleError({
+            swapError: {
+              code: 'USER_REJECTED',
+              error: 'User rejected the request.',
+            },
+          });
+        } else {
+          handleError({ swapError: e as SwapError });
+        }
       } finally {
         setLoading(false);
       }
     },
-    [address, handleError, from.amount, from.token, to.token],
+    [
+      address,
+      config,
+      handleError,
+      from.amount,
+      from.token,
+      sendTransactionAsync,
```

Comment:
What's `sendTransactionAsync`?

Reply:
https://wagmi.sh/react/api/hooks/useSendTransaction#mutate-async

so we can `await` it (to use the transaction hash)